src/elgamal.rs

```diff
@@ -0,0 +1,244 @@
+            let prod = Integer::from(&i_code * &hy);
+            let d = prod.div_rem_euc_ref(&key.p).complete().1;
+            // add the pair to the cipher pairs list
+            let mut arr: Vec<Integer> = Vec::new();
+            arr.push(c);
+            arr.push(d);
+            cipher_pairs.push(arr);
+        }
+        let mut encrypted_str = "".to_string();
+        for pair in cipher_pairs {
+            let pair_one = pair[0].to_string_radix(STR_RADIX).to_string();
+            let pair_two = pair[1].to_string_radix(STR_RADIX).to_string();
+            let space = " ".to_string();
+
+            encrypted_str += &pair_one;
+            encrypted_str += &space;
+            encrypted_str += &pair_two;
+            encrypted_str += &space;
+        }
+        encrypted_str
+    }
+
+    ///Performs decryption on the cipher pairs found in Cipher using
+    ///private key K2 and writes the decrypted values to file Plaintext.
+    fn decrypt(&self,key: &PrivateKey) -> Option<String> {
+        // check if the last char is space
+        let mut cipher_chars = self.chars();
+        if let Some(last) = cipher_chars.clone().last() {
+            if last.is_whitespace() {
+                // if the last char is space, removed it from the string.
+                cipher_chars.next_back();
+            }
+        } else {
+            // if the cipher string is empty, return None.
+            return None;
+        }
+        let reduced_str = cipher_chars.as_str();
+        let ciphers = reduced_str.split(" ").collect::<Vec<&str>>();
+        let count = ciphers.len();
+        if count % 2 != 0 {
+            return None;
+        }
+        let mut plain_text = Vec::new();
+        for cd in ciphers.chunks(2) {
+            // c = first number in pair
+            let c = cd[0];
+            let c_int = Integer::from_str_radix(c, STR_RADIX).unwrap();
+            // d = second number in pair
+            let d = cd[1];
+            let d_int = Integer::from_str_radix(d, STR_RADIX).unwrap();
+            // s = c^x mod p
+            let s = c_int.pow_mod(&key.x, &key.p).unwrap();
+            // plaintext integer = ds^-1 mod p
+            let p_2 = Integer::from(&key.p - 2);
+            let mod_exp_s = s.pow_mod(&p_2, &key.p).unwrap();
+            let d_by_mod = Integer::from(d_int * mod_exp_s);
+            let plain_i = Integer::from(d_by_mod.div_rem_euc_ref(&key.p).complete().1);
+            // add plain to list of plaintext integers
+            plain_text.push(plain_i);
+            // count the length of the cipher strings
+        }
+        Some(decode_utf16(&plain_text, key.bit_length.clone()))
+    }
+}
+
+///Encodes bytes to integers mod p.
+///         Example
+///         if n = 24, k = n / 8 = 3
+///         z[0] = (summation from i = 0 to i = k)m[i]*(2^(8*i))
+///         where m[i] is the ith message byte
+pub fn encode_utf16(s_plaintext: &str, bit_length: u32) -> Vec<Integer> {
```

Comment:
if there is no need for external reference, we may make it private method. (remove pub)